File: scala/src/main/scala/fun/pierre/lists/P03.scala
package fun.pierre.lists

import java.util.NoSuchElementException

import scala.annotation.tailrec

object P03 {

  @tailrec
  def kThElement[T](list: List[T], k: Int, counter: Int = 1): T = counter match {
    case `k` => list.head
    case _ if list.nonEmpty => kThElement(list.drop(1), k, counter + 1)
    case _ => throw new NoSuchElementException("k past list length")
  }
}

File: java/src/main/java/fun/pierre/lists/P03.java
package fun.pierre.lists;

import java.util.List;

public class P03 {
    public static<T> T kThElement(List<T> _list, int _k) {
        int size = _list.size();
        if (size <= _k) {
            throw new IllegalArgumentException(String.format("List has less than %d elements", _k));
        }

        return _list.get(_k - 1);
    }
}
